File: codes/matlab/ch02/wave_equation_waterfall.m
%% wave_equation_waterfall.m
%
% Waterfall (3D surface) plot showing the complete space-time evolution of the
% wave equation solution. This visualization demonstrates the standing wave
% oscillations of a plucked string over time.
%
% Wave equation (Dirichlet BCs on [0, L]):
%     u_tt = c^2 u_xx
%
% Analytical solution:
%     u(x,t) = sum (a_n cos(omega_n t) + b_n sin(omega_n t)) sin(n*pi*x/L)
%
% where omega_n = c*n*pi/L
%
% Plucked string initial condition (triangle):
%     f(x) = (2h/L)*x           for 0 <= x <= L/2
%     f(x) = 2h*(1 - x/L)       for L/2 <= x <= L
%     g(x) = 0                  (initially at rest)
%
% Author: Dr. Denys Dutykh
%         Mathematics Department
%         Khalifa University of Science and Technology
%         Abu Dhabi, UAE
%
% Part of the book "Computational Études: A Spectral Approach"
% https://github.com/dutykh/computational-etudes

clear; clc; close all;

%% Configuration
N_MODES = 50;           % Number of Fourier modes in truncation
NX = 150;               % Spatial grid points
NT = 100;               % Time grid points
L = pi;                 % Domain length [0, L]
C = 1.0;                % Wave speed
H = 1.0;                % Height of plucked string at center

% Period of oscillation: T = 2L/c
T_PERIOD = 2 * L / C;
T_MAX = T_PERIOD;       % Show one full period

% Output path (relative to this script's location)
script_dir = fileparts(mfilename('fullpath'));
output_file = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch02', 'matlab', 'wave_waterfall.pdf');

%% Compute Fourier coefficients for the plucked string
a_n = zeros(1, N_MODES + 1);
b_n = zeros(1, N_MODES + 1);

for n = 1:N_MODES
    if mod(n, 2) == 1  % odd n only
        if mod(n, 4) == 1
            sgn = 1;
        else
            sgn = -1;
        end
        a_n(n + 1) = sgn * 8.0 * H / (n^2 * pi^2);
    end
end

%% Create 2D grid for space-time
x = linspace(0, L, NX);
t = linspace(0, T_MAX, NT);
[X, T] = meshgrid(x, t);

%% Evaluate solution on grid
U = zeros(size(X));
for i = 1:NT
    ti = t(i);
    u = zeros(1, NX);
    for n = 1:N_MODES
        omega_n = C * n * pi / L;
        spatial = sin(n * pi * x / L);
        temporal = a_n(n + 1) * cos(omega_n * ti) + b_n(n + 1) * sin(omega_n * ti);
        u = u + temporal * spatial;
    end
    U(i, :) = u;
end

%% Create figure
fig = figure('Units', 'inches', 'Position', [1, 1, 7, 5]);

% Custom colormap: navy -> white -> sky blue
n_colors = 256;
navy_rgb = [20, 45, 110] / 255;
white_rgb = [1, 1, 1];
sky_rgb = [120, 150, 210] / 255;

cmap = zeros(n_colors, 3);
for i = 1:n_colors/2
    t_val = (i - 1) / (n_colors/2 - 1);
    cmap(i, :) = navy_rgb * (1 - t_val) + white_rgb * t_val;
end
for i = 1:n_colors/2
    t_val = (i - 1) / (n_colors/2 - 1);
    cmap(n_colors/2 + i, :) = white_rgb * (1 - t_val) + sky_rgb * t_val;
end
colormap(cmap);

% Plot the surface
surf(X, T, U, 'EdgeColor', [0.5, 0.5, 0.5], 'EdgeAlpha', 0.3, ...
     'FaceAlpha', 0.9, 'FaceLighting', 'gouraud');
hold on;

% Labels
xlabel('$x$', 'Interpreter', 'latex', 'FontSize', 11);
ylabel('$t$', 'Interpreter', 'latex', 'FontSize', 11);
zlabel('$u(x,t)$', 'Interpreter', 'latex', 'FontSize', 11);

% Custom x-ticks with pi notation
set(gca, 'XTick', [0, L/4, L/2, 3*L/4, L]);
set(gca, 'XTickLabel', {'$0$', '$\pi/4$', '$\pi/2$', '$3\pi/4$', '$\pi$'}, ...
    'TickLabelInterpreter', 'latex');

% Custom t-ticks as fractions of period
set(gca, 'YTick', [0, T_PERIOD/4, T_PERIOD/2, 3*T_PERIOD/4, T_PERIOD]);
set(gca, 'YTickLabel', {'$0$', '$T/4$', '$T/2$', '$3T/4$', '$T$'}, ...
    'TickLabelInterpreter', 'latex');

% Set viewing angle
view(-60, 25);

% Add lighting for better 3D perception
camlight('headlight');
lighting gouraud;

%% Save figure
% Ensure output directory exists
[output_dir, ~, ~] = fileparts(output_file);
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

% Export to PDF
exportgraphics(fig, output_file, 'ContentType', 'vector', 'Resolution', 300);
fprintf('Figure saved to: %s\n', output_file);

close(fig);

File: codes/julia/ch02/wave_equation_waterfall.jl
# wave_equation_waterfall.jl
#
# Waterfall (3D surface) plot showing the complete space-time evolution of the
# wave equation solution. This visualization demonstrates the standing wave
# oscillations of a plucked string over time.
#
# Wave equation (Dirichlet BCs on [0, L]):
#     u_tt = c^2 u_xx
#
# Analytical solution:
#     u(x,t) = Sum (a_n cos(w_n t) + b_n sin(w_n t)) sin(n pi x / L)
#
# where w_n = c n pi / L
#
# Plucked string initial condition (triangle):
#     f(x) = (2h/L)x           for 0 <= x <= L/2
#     f(x) = 2h(1 - x/L)      for L/2 <= x <= L
#     g(x) = 0                 (initially at rest)
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const N_MODES  = 50    # Number of Fourier modes in truncation
const NX       = 150   # Spatial grid points
const NT       = 100   # Time grid points
const L_DOMAIN = Float64(pi)  # Domain length [0, L]
const C_SPEED  = 1.0   # Wave speed
const H_HEIGHT = 1.0   # Height of plucked string at center

# Period of oscillation: T = 2L/c
const T_PERIOD = 2 * L_DOMAIN / C_SPEED
const T_MAX    = T_PERIOD  # Show one full period

# Book color scheme
const NAVY = colorant"#142D6E"
const SKY  = colorant"#7896D2"

# Output path (relative to this script's location)
SCRIPT_DIR = @__DIR__
OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch02", "julia")
mkpath(OUTPUT_DIR)

# -----------------------------------------------------------------------------
# Fourier coefficients for the plucked string
# -----------------------------------------------------------------------------
"""
    plucked_string_coefficients(n_modes, L, h)

Compute Fourier sine coefficients for the plucked string initial condition.
"""
function plucked_string_coefficients(n_modes, L, h)
    a_n = zeros(n_modes + 1)
    b_n = zeros(n_modes + 1)

    for n in 1:n_modes
        if n % 2 == 1  # odd n only
            sign_val = (n % 4 == 1) ? 1 : -1
            a_n[n+1] = sign_val * 8.0 * h / (n^2 * pi^2)
        end
    end

    return a_n, b_n
end

# -----------------------------------------------------------------------------
# Evaluate the wave equation solution
# -----------------------------------------------------------------------------
"""
    wave_solution(x, t, a_n, b_n, L, c)

Evaluate the wave equation solution at positions `x` and time `t`.

    u(x,t) = Sum_{n=1}^N (a_n cos(w_n t) + b_n sin(w_n t)) sin(n pi x / L)
"""
function wave_solution(x, t, a_n, b_n, L, c)
    u = zeros(length(x))

    n_modes = length(a_n) - 1
    for n in 1:n_modes
        omega_n = c * n * pi / L
        spatial  = @. sin(n * pi * x / L)
        temporal = a_n[n+1] * cos(omega_n * t) + b_n[n+1] * sin(omega_n * t)
        @. u += temporal * spatial
    end

    return u
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Compute Fourier coefficients
    a_n, b_n = plucked_string_coefficients(N_MODES, L_DOMAIN, H_HEIGHT)

    # Create 2D grid for space-time
    x = range(0, L_DOMAIN, length=NX)
    t = range(0, T_MAX, length=NT)

    # Evaluate solution on grid
    U = zeros(NT, NX)
    for i in 1:NT
        U[i, :] = wave_solution(collect(x), t[i], a_n, b_n, L_DOMAIN, C_SPEED)
    end

    # Create custom colormap (navy to white to sky)
    navy_rgb  = (20/255, 45/255, 110/255)
    white_rgb = (1.0, 1.0, 1.0)
    sky_rgb   = (120/255, 150/255, 210/255)

    n_colors = 256
    colors_list = Vector{RGBf}(undef, n_colors)
    for i in 1:n_colors÷2
        t_val = (i - 1) / (n_colors÷2 - 1)
        colors_list[i] = RGBf(
            navy_rgb[1] * (1 - t_val) + white_rgb[1] * t_val,
            navy_rgb[2] * (1 - t_val) + white_rgb[2] * t_val,
            navy_rgb[3] * (1 - t_val) + white_rgb[3] * t_val
        )
    end
    for i in 1:n_colors÷2
        t_val = (i - 1) / (n_colors÷2 - 1)
        colors_list[n_colors÷2 + i] = RGBf(
            white_rgb[1] * (1 - t_val) + sky_rgb[1] * t_val,
            white_rgb[2] * (1 - t_val) + sky_rgb[2] * t_val,
            white_rgb[3] * (1 - t_val) + sky_rgb[3] * t_val
        )
    end
    custom_cmap = cgrad(colors_list)

    # Create figure with 3D surface
    fig = Figure(size = (700, 500))
    ax = Axis3(fig[1, 1],
               xlabel = L"x", ylabel = L"t", zlabel = L"u(x,t)",
               xlabeloffset = 40, ylabeloffset = 40, zlabeloffset = 50,
               xticklabelsize = 9, yticklabelsize = 9, zticklabelsize = 9,
               azimuth = -60 * pi / 180 + pi,
               elevation = 25 * pi / 180)

    # Plot the surface
    surface!(ax, collect(x), collect(t), U',
             colormap = custom_cmap, alpha = 0.9)

    # Add wireframe for better depth perception
    wireframe!(ax, collect(x), collect(t), U',
               color = RGBAf(0.5, 0.5, 0.5, 0.3), linewidth = 0.2,
               overdraw = false)

    # Custom x-ticks with pi notation
    ax.xticks = ([0, L_DOMAIN/4, L_DOMAIN/2, 3*L_DOMAIN/4, L_DOMAIN],
                 [L"0", L"\frac{\pi}{4}", L"\frac{\pi}{2}", L"\frac{3\pi}{4}", L"\pi"])

    # Custom t-ticks as fractions of period
    ax.yticks = ([0, T_PERIOD/4, T_PERIOD/2, 3*T_PERIOD/4, T_PERIOD],
                 [L"0", L"\frac{T}{4}", L"\frac{T}{2}", L"\frac{3T}{4}", L"T"])

    # Save figures
    save(joinpath(OUTPUT_DIR, "wave_waterfall.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "wave_waterfall.png"), fig, px_per_unit = 4)
    println("Figure saved to: $(joinpath(OUTPUT_DIR, "wave_waterfall.pdf"))")
end

main()
